x/poolmanager/README.md

```diff
@@ -330,3 +330,52 @@ higher slippage.
+
+The `EstimateTradeBasedOnPriceImpact` query allows users to estimate a trade for all pool types given the following parameters are provided for this request `EstimateTradeBasedOnPriceImpactRequest`:
+
+
+- **FromCoin**: (`sdk.Coin`): the total amount of tokens one wants to sell.
+- **ToCoinDenom**: (`string`): is the denom they want to buy with the tokens being sold.
+- **PoolId**: (`uint64`): is the identifier of the pool that the trade will happen on.
+- **MaxPriceImpact**: (`sdk.Dec`): is the maximum percentage that the user is willing to affect the price of the pool.
+- **TwapPrice**: (`sdk.Dec`) is an external price that the user can optionally enter to have the `MaxPriceImpact` adjusted as the `SpotPrice` of a pool could be changed at any time.
+
+The response would be `EstimateTradeBasedOnPriceImpactResponse` which contains the following data:
+
+- **InputCoin**: (`sdk.Coin`): the actual input amount that would be tradeable under that price impact (might be the full amount).
+- **OutputCoin**: (`sdk.Coin`): the amount of the `ToCoinDenom` tokens being received for the actual `InputCoin` trade.
+
+With that data it is easier for any entity to fill in the `MsgSwapExactAmountIn` details.
+
+### Process
+
+The following is the process in which the query finds a trade that will stay below the `MaxPriceImpact` value.
+
+1. Verify `PoolId`, `FromCoin`, `ToCoinDenom` are not empty
+2. Return the specific `swapModule` based on the `PoolId`
+3. Return the specific `PoolI` interface from the `swapModule` based on the `PoolId`
+4. Calculate the `SpotPrice` in terms of the token being bought, therefore if it's an OSMO/ATOM pool and OSMO is being sold we need to calculate the `SpotPrice` in terms of `ATOM` being the base asset and OSMO being the quote asset.
+5. If we have a `TwapPrice` specified in the request we need to adjust the `MaxPriceImpact` into a new variable `adjustedMaxPriceImpact` which would either increase if the `SpotPrice` is cheaper than the `TwapPrice` or decrease if the `SpotPrice` is more expensive leaving less room to estimate a trade.
+   1. If the `adjustedMaxPriceImpact` was calculated to be `0` or negative it means that the `SpotPrice` is more expensive than the `TwapPrice` and has already exceeded the possible `MaxPriceImpact`. We return a `sdk.ZeroInt()` input and output for the input and output coins indicating that no trade is viable.
+6. Firstly we execute `CalcOutAmtGivenIn` using the initial amount given from `FromCoin` without a swap fee as we're only interested in the price impact of the trade which doesn't involve swap fees.
+   1. If this errors there is very little liquidity for the pool to estimate anything.
+   2. If the amount of tokens returned is `0` it means we've entered a very small amount and we are trying to estimate a trade on dust. We should return the full input amount `FromCoin` as the `InputCoin` and `OutputCoin` of `sdk.ZeroInt()` to indicate we will get `0` from our trade.
```

Comment:
Is there a scenario where the pool has low liquidity but doesn't error, and we still get 0 returned?
(i.e. option 2 but with very little liquidity)

In this case we would be marking a non-dust amount as dust.